File: logic/field.py
"""Gaussian-field priority score used by the greedy edge-growth algorithm."""
import math
from typing import Any

PRIORITY_OFFSET: float = 0.0


def gaussian(x: tuple[float, float], mu: tuple[float, float], sigma: float) -> float:
    """Unnormalized isotropic 2D Gaussian bump centred at mu."""
    squaredDistance = (x[0] - mu[0]) ** 2 + (x[1] - mu[1]) ** 2
    return math.exp(-squaredDistance / (2 * sigma * sigma))


def fieldValue(dsu: Any, vertex: Any, x: tuple[float, float], sigma: float) -> float:
    """Strength-scaled Gaussian bump centred on `vertex` itself, evaluated at
    point x. The strength is sqrt of the special-subset count of vertex's current component.
    """
    specialCount = dsu.sCount(dsu.find(vertex))
    if specialCount == 0:
        return 0.0

    return math.sqrt(specialCount) * gaussian(x, vertex, sigma)


def key(dsu: Any, u: Any, v: Any, sigma: float) -> float:
    """Symmetric priority score for the unordered pair {u, v}."""
    fieldOfVAtU = fieldValue(dsu, v, u, sigma)
    fieldOfUAtV = fieldValue(dsu, u, v, sigma)
    return PRIORITY_OFFSET + fieldOfVAtU + fieldOfUAtV

File: logic/field.js
/** Gaussian-field priority score used by the greedy edge-growth algorithm. */
import { gaussian } from "./rng.js";

// CONSTANTS

const DAMPENING_FACTOR = 0.1;
const STRENGTHENING_FACTOR = 10.0;
export const SIGMA_DIVISOR = 15.0;

// HELPER FUNCTIONS - STRENGTH CALCULATION

function _strength(dsu, vertex, x) {
  const dampeningFactor = dsu.connected(vertex, x) ? DAMPENING_FACTOR : 1;
  return Math.sqrt(dampeningFactor * (dsu.componentSize(vertex)));
}

// PUBLIC INTERFACE

/** Derives the Gaussian spread sigma from the grid size. */
export function sigmaFromGridSize(gridSize) {
  return gridSize / SIGMA_DIVISOR;
}

/** Strength-scaled Gaussian bump centred on `vertex`, evaluated at x. */
export function fieldValue(dsu, vertex, x, sigma) {
  const strengtheningFactor = dsu.isSpecial(vertex) ? STRENGTHENING_FACTOR : 1.0;
  return strengtheningFactor * _strength(dsu, vertex, x) * gaussian(x, vertex, sigma);
}

/** Symmetric priority score for the unordered pair {u, v}. */
export function key(dsu, u, v, sigma) {
  const fieldOfVAtU = fieldValue(dsu, v, u, sigma);
  const fieldOfUAtV = fieldValue(dsu, u, v, sigma);
  return fieldOfVAtU + fieldOfUAtV;
}
